File: Java/62.unique_paths.java
// Date: 11-26-2019
// Tag: Dynamic Programming/Math

class Solution {
    public int uniquePaths(int m, int n) {
        if (m == 0 || n == 0) return 0;
        int[][] dp = new int[m][n];

        dp[0][0] = 1;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (valid(i - 1, j, dp)) dp[i][j] += dp[i - 1][j];
                if (valid(i, j - 1, dp)) dp[i][j] += dp[i][j - 1];
            }
        }

        return dp[m - 1][n - 1];
    }

    public boolean valid(int row, int col, int[][] dp) {
        if (row < 0 || row >= dp.length || col < 0 || col >= dp[0].length) return false;
        return true;
    }
}

// Math Formula -> Combination(N, K) for referrence
public int uniquePaths(int m, int n) {
    double value = 1;
    // here we calculate the total possible path number
    // Combination(N, k) = n! / (k!(n - k)!)
    // reduce the numerator and denominator and get
    // C = ( (n - k + 1) * (n - k + 2) * ... * n ) / k!
    for (int i = 1; i <= n - 1; i++) {
        value *= ((double) (m + i - 1) / (double) i);
    }
    return (int) Math.round(value);
}

File: Leetcode_solutions_by_company/Goldman Sachs/Frontend Role Preparation/62.unique_paths.js
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths = function(m, n) {
    if (m == 0 || n == 0 || m == null || n == null) return 0;

    let dp = new Array(m);
    for (let i = 0; i < m; i++) {
        dp[i] = new Array(n).fill(1);
    }


    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
        }
    }

    return dp[m - 1][n - 1];
};

public class Solution {
    public int uniquePaths(int m, int n) {
    int[] arr = new int[m];
    for (int i = 0; i < m; i++) {
    arr[i] = 1;
}
for (int i = 1; i < n; i++) {
    for (int j = 1; j < m; j++) {
        arr[j] = arr[j] + arr[j-1];
    }
}
return arr[m-1];
}
}